### PR TITLE
#53: Implement multiple commands for un-/re-/loading cogs

### LINKED ISSUE #53: Commands for loading, unloading or reloading a Cog
The command should be inside the Admin Cog and can therefore only be invoked by the server owner.

This way we can deactivate specific kinds of commands or introduce new ones without taking the bot offline. Once the extension reloads, any changes that we did will be applied.

### PR DESCRIPTION
**Pull request by PKlempe**
Closes #53

### CODE REVIEW

File: bot/admin/admin.py [hunk 1]
```diff
@@ -133,6 +133,85 @@ async def embed_available_cogs(self, ctx: commands.Context):
...
+
+        Loads an extension (Cog) with the specified name into the bot.
+
+        Args:
+            ctx (discord.ext.commands.Context): The context from which this command is invoked.
+            extn_name (str): The name of the extension (Cog).
+        """
+        extn_name = extn_name.capitalize()
+
+        if "Cog" and "cog" not in extn_name:
```
Comment (SebastiantheAstorian) [inline bot/admin/admin.py:149]: I guess the name should end with cog? 
In that case use function  [endswith() ](https://www.tutorialspoint.com/python3/string_endswith.htm)

Comment (SebastiantheAstorian) [inline bot/admin/admin.py:149]: Also this check is completely wrong I think. IMO it evaluates: 
`"Cog" == True && "cog" not in extn_name`

So the first "Cog" is only evaluated if it is true, and a non-empty string is always true.

Comment (SebastiantheAstorian) [reply bot/admin/admin.py:149]: Btw. thats the stuff you should test in a test. If it correctly evaluates and adds the strings (therefor it also would fit fine in a private function)

File: bot/admin/admin.py [hunk 2]
```diff
@@ -133,6 +133,85 @@ async def embed_available_cogs(self, ctx: commands.Context):
...
+            extn_name += "Cog"
+        elif "cog" in extn_name:
+            extn_name = extn_name[:-3] + "Cog"
+
+        self.bot.load_extension(constants.INITIAL_EXTNS[extn_name])
```
Comment (SebastiantheAstorian) [inline bot/admin/admin.py:154]: Have you tested what happens if you enter an invalid cog name?
!bot load yolocog

File: bot/admin/admin.py [hunk 3]
```diff
@@ -133,6 +133,85 @@ async def embed_available_cogs(self, ctx: commands.Context):
...
+
+        Removes an extension (Cog) with the specified name from the bot.
+
+        Args:
+            ctx (discord.ext.commands.Context): The context from which this command is invoked.
+            extn_name (str): The name of the extension (Cog).
+        """
+        extn_name = extn_name.capitalize()
+
+        if "Cog" and "cog" not in extn_name:
```
Comment (SebastiantheAstorian) [inline bot/admin/admin.py:170]: Same as above in load function. This does not work and is highly unstable

Comment (SebastiantheAstorian) [reply bot/admin/admin.py:170]: Also if you had this code where you replace the name in a separate function, not only could you test it beforehand with a test, you could also reuse it and not duplicate the same code.

File: bot/admin/admin.py [hunk 4]
```diff
@@ -133,6 +133,85 @@ async def embed_available_cogs(self, ctx: commands.Context):
...
+        Reloads an extension (Cog) with the specified name from the bot. If changes to the code inside a Cog have been
+        made, this is going to apply them without taking the bot offline.
+
+        Args:
+            ctx (discord.ext.commands.Context): The context from which this command is invoked.
+            extn_name (str): The name of the extension (Cog).
+        """
+        extn_name = extn_name.capitalize()
+
+        if "Cog" and "cog" not in extn_name:
```
Comment (SebastiantheAstorian) [inline bot/admin/admin.py:192]: Same as above. 

Also third time copying code...

File: bot/admin/admin.py [hunk 5]
```diff
@@ -133,6 +133,85 @@ async def embed_available_cogs(self, ctx: commands.Context):
...
+        """Command handler for the `bot reload` subcommand `all`.
+
+        Reloads all the extension (Cogs) from the bot. If changes to the code inside a Cog have been
+        made, this is going to apply them without taking the bot offline.
+
+        Args:
+            ctx (discord.ext.commands.Context): The context from which this command is invoked.
+        """
+        for extn in constants.INITIAL_EXTNS:
+            self.bot.reload_extension(constants.INITIAL_EXTNS[extn])
```
Comment (SebastiantheAstorian) [inline bot/admin/admin.py:212]: Why do you access the dictionary, using the dictionary?
Wouldn't it be simpler to just:

```
for extn in constants.INITIAL_EXTNS.values():
    self.bot.reload_extension(etxn)
```

alternatively in a more hacky way:  
```
map(lambda extn: self.bot.reload_extension(extn), constants.  INITIAL_EXTNS.values())
```
or 
```
[ self.bot.reload_extension(extn) for extn in constants.INITIAL_EXTNS.values()]
```

File: bot/admin/admin.py (hunk 1 above)
Comment (PKlempe) [reply bot/admin/admin.py:149]: You're completely right. It seems that I didn't have enough caffeine this morning and my brain probably thought that this makes sense in the way you read this. 😂

File: bot/admin/admin.py (hunk 2 above)
Comment (PKlempe) [reply bot/admin/admin.py:154]: Throws an error like "InvalidCogName". I'm gonna make the bot inform the user about this.

File: bot/admin/admin.py (hunk 4 above)
Comment (PKlempe) [reply bot/admin/admin.py:192]: garbage

Comment (PKlempe) [reply bot/admin/admin.py:192]: But yeah, I was in a hurry. I'm gonna right a function for this. 😅

**Pull request merged**